src/decision/engine.js

```diff
@@ -0,0 +1,76 @@
+'use strict'
+
+const debug = require('debug')
+const log = debug('engine')
+log.error = debug('engine:error')
+
+const Wantlist = require('../wantlist')
+const PeerRequestQueue = require('./peer-request-queue')
+const Ledger = require('./ledger')
+
+module.exports = class Engine {
+  constructor (blockStore) {
+    this.blockStore = blockStore
+
+    // A list of of ledgers by their partner id
+    this.ledgerMap = new Map()
+
+    // A priority queue of requests received from different
+    // peers.
+    this.peerRequestQueue = new PeerRequestQueue()
+  }
+
+  messageReceived (peerId, msg) {
+    if (msg.empty) {
+      log('received empty message from %s', peerId)
+    }
+
+    const ledger = this._findOrCreate(peerId)
+
+    // If the message was a full wantlist clear the current one
+    if (msg.full) {
+      ledger.wantlist = new Wantlist()
+    }
+
+    for (let entry of msg.wantlist.values()) {
+      if (entry.cancel) {
+        log('cancel %s', entry.key)
+        ledger.cancelWant(entry.key)
+        this.peerRequestQueue.remove(entry.key, peerId)
+      } else {
+        log('wants %s - %s', entry.key, entry.priority)
+        ledger.wants(entry.key, entry.priority)
+
+        // If we already have the block, serve it
+        if (this.blockStore.has(entry.key)) {
+          this.peerRequestQueue.push(entry.entry, peerId)
```

Comment:
Is this serving the block or just adding to the list of things that the peer wants?